controller/controlplane/controller_reconciler_utils.go

```diff
@@ -764,3 +797,70 @@ func (r *Reconciler) ensureValidatingWebhookConfiguration(
 
 	return op.Created, r.Client.Create(ctx, generatedWebhookConfiguration)
 }
+
+func (r *Reconciler) validateReferenceGrants(
+	ctx context.Context,
+	cp *operatorv1beta1.ControlPlane,
+) error {
+	if cp.Spec.WatchNamespaces == nil {
+		return nil
+	}
+
+	switch cp.Spec.WatchNamespaces.Type {
+	// NOTE: We currentlty do not require any ReferenceGrants or other permission
+	// granting resources for the "All" case.
+	case operatorv1beta1.WatchNamespacesTypeAll:
+		return nil
+	// No special permissions are required to watch the controlplane's own namespace.
+	case operatorv1beta1.WatchNamespacesTypeOwn:
+		return nil
+	case operatorv1beta1.WatchNamespacesTypeList:
+		for _, ns := range cp.Spec.WatchNamespaces.List {
+			if err := ensureReferenceGrantsForNamespace(ctx, r.Client, cp, ns); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	default:
+		return nil
+	}
+}
+
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=referencegrants,verbs=list
+
+// ensureReferenceGrantsForNamespace ensures that a ReferenceGrant exists for the
+// given namespace and ControlPlane.
+// It returns an error if a ReferenceGrant is missing.
+func ensureReferenceGrantsForNamespace(
+	ctx context.Context,
+	cl client.Client,
+	cp *operatorv1beta1.ControlPlane,
+	ns string,
+) error {
+	var refGrants gatewayv1beta1.ReferenceGrantList
+	if err := cl.List(ctx, &refGrants, client.InNamespace(ns)); err != nil {
+		return fmt.Errorf("failed listing ReferenceGrants in namespace %s: %w", ns, err)
+	}
+	for _, refGrant := range refGrants.Items {
+		if !lo.ContainsBy(refGrant.Spec.From, func(from gatewayv1beta1.ReferenceGrantFrom) bool {
+			return from.Group == "gateway-operator.konghq.com" &&
```

Comment:
Could we use a group extracted from the package variable instead of the hardcoded constant?

Reply:
Updated, PTAL.